server/src/main/java/org/apache/cassandra/sidecar/job/OperationsJob.java

```diff
@@ -0,0 +1,152 @@
+    {
+        this.jobId = jobId;
+        this.status = status;
+        this.failureReason = "";
+    }
+
+    public void setStatus(OperationsJobStatus status)
+    {
+        this.status = status;
+    }
+    public OperationsJobStatus status()
+    {
+        return status;
+    }
+    public String failureReason()
+    {
+        return failureReason;
+    }
+
+    public UUID jobId()
+    {
+        return jobId;
+    }
+
+    /**
+     * Supplier specifying the functionality of the job to be triggered when the job is executed. Subclasses to
+     * provide operation-specific implementations
+     * @return a function with the operation implementation that returns a {@code JobResult}
+     * @throws Exception
+     */
+    public abstract Supplier<OperationsJobResult> jobOperationSupplier() throws Exception;
+
+    /**
+     * Provide a meaningful name of the operation executed by the concrete subclass.
+     * @return the name of the operation. eg. nodetool command name
+     */
+    public abstract String operation();
+
+    /**
+     * Specifies the downstream operation to be performed to check if the job is running on the Cassandra node/cluster.
+     * This functionality is provided by the Job when it is asynchronously triggering the job via the {@code JobManager}
+     * For synchronous jobs this should always return false.
+     * @return true if the job is running downstream
+     */
+    public abstract boolean checkInflightJob();
+
+    /**
+     * Execute the job behavior as specified in the operation supplier {@link #jobOperationSupplier()},
+     * while tracking the status of the job's lifecycle.
+     */
+    public void execute()
+    {
+        try
+        {
+            LOGGER.info("Executing job with ID: {}", jobId);
+            OperationsJobResult result = jobOperationSupplier().get();
+            status = result.status();
+            failureReason = result.reason();
+            LOGGER.debug("Job with ID: {} returned with status: {}", jobId, status);
+        }
+        catch (Exception e)
+        {
+            String reason = (e.getCause() != null) ? e.getCause().getMessage() : e.getMessage();
+            LOGGER.error("Failed to execute job {} with reason: {}", jobId, reason);
+            status = OperationsJobStatus.Failed;
+            failureReason = reason;
+        }
+        finally
+        {
+            latch.countDown();
+        }
+    }
+
+    /**
+     * Returns if the job execution completes within the provided wait time (in seconds)
+     * @param waitSeconds no. of seconds to wait for the job execution to complete
+     * @return true if the job completed within the specified time
+     */
+    public boolean isResultAvailable(long waitSeconds)
+    {
+        try
+        {
+            return (!latch.await(waitSeconds, TimeUnit.SECONDS)) ? false : true;
+        }
+        catch (final InterruptedException e)
+        {
+            return false;
+        }
+    }
```

Comment:
Not at a fan of this blocking API.

Consider this option. Schedule a future task / timer, and check the result after timer times out. 

```
    public boolean isComplete()
    {
        // of course, not necessary to use CountDownLatch in this case
        return latch.getCount() == 0;
    }

    // at call-site
    Vertx vertx;
    OperationsJob operationsJob;
    vertx.timer(waitSeconds, TimeUnit.SECONDS).map(v -> operationsJob.isComplete())...
```

Reply:
I see the concern, but in this case, we will have - 
1) handler of long-running job invoking `trySubmitJob` within `executorPools.service().executeBlocking(()`
2) Job execution/submission done within `executorPools.internal().runBlocking()`

So we have a multi-level indirection. For vertx to wait for the execution to complete, we would need some form of Future chaining to detect completion of the inner most execution thread. 

We will have to weigh the complexity of that to the existing implementation. Thoughts?